### DISCUSSION TITLE
"Official" Website - Opinions/suggestions?

### DISCUSSION THREAD

**Discussion by drkameleon** [arturo-lang/arturo#179] [Show and tell]
After a long time thinking what to do with the website, I finally did a makeover.

There are some issues pending, but I think I'm pretty satisfied if we consider it a first... "working draft".

https://arturo-lang.io

<img width="1427" alt="Screenshot 2021-02-16 at 13 56 12" src="https://user-images.githubusercontent.com/1265028/108065765-c3413780-705e-11eb-8057-6686af5a23b5.png">

Some of the pending things:

- Add a "try it online" console/REPL (most likely with Ace editor - anything else to suggest?)
- Add a "Function cheatsheet" (side-by-side with the online REPL)
- Also have a website version of the documentation? (@https://github.com/arturo-lang/arturo/wiki)

I don't know. Tell me what you think about it! 100% open to ideas/feedback/suggestions!

**Comment by celtic-coder** [comment]
Hi Yanis (@drkameleon),

I ran a desktop application that check the links on your site and found the following (somewhat minor) problems:

1. The link https://fonts.gstatic.com is, I'm guessing, found on every page of your site. However, this link is broken and returns a 404 error. Is this link required?

![fonts-link-in-html](https://user-images.githubusercontent.com/11787866/142781820-d353866d-9396-4042-a82a-f5fdb5818348.png)

By the way, if you intend to use Google Fonts, you might want to check out this article on [Google Fonts and privacy](https://www.brycewray.com/posts/2020/08/google-fonts-privacy//). It links to an [issue](https://github.com/google/fonts/issues/1495) on the Google Fonts GitHub repo about GDPR compliance. It seems the resolution is to host the fonts locally.

2. The Creative Commons link (http://creativecommons.org/licenses/by-nc-sa/4.0/) is now a secure "https" link (https://creativecommons.org/licenses/by-nc-sa/4.0/).

3. The same is true of the MIT License (http://opensource.org/licenses/mit-license.php) which is permanently redirected to (https://opensource.org/licenses/mit-license.php).

4. The following two links have the same meta description (calculate the modulo given values and return result - Arturo programming). Is this intentional?

* https://arturo-lang.io/documentation/library/arithmetic/mod/
* https://arturo-lang.io/documentation/library/arithmetic/mul/

5. Similarly, the following links have the same meta description (flatten given collection by eliminating nested blocks - Arturo programming):

* https://arturo-lang.io/documentation/library/iterators/fold/
* https://arturo-lang.io/documentation/library/collections/flatten/

6. Finally, the same problem occurs in these links (return the logical NAND for the given values - Arturo programming):

* https://arturo-lang.io/documentation/library/logic/nor-/
* https://arturo-lang.io/documentation/library/logic/nand-/

From looking at the HTML of your site - particularly the documentation portion - it looks to be autogenerated from an underlying database. Is this correct? If so, this is quite an achievement.

Hmm, I guess if I had looked further, I would have answered my own question! 😄 The [wiki](https://github.com/arturo-lang/arturo/wiki) mentions that the documentation is generated with "docgen.art".  Good work!

I had looked into Rebol/Red a few years back and only in the past few months stumbled on Forth and related languages. Arturo looks to be an interesting combination! I'm looking forward to getting into the language.

Kind Regards,
Liam

**Reply by drkameleon** [reply]
First of all, thanks a lot for the feedback.

I've done tooooo much web dev these days (hence, my virtual absence from the open source scene in the past few months - actually) so I can totally see your point.

I'll fix all of the HTML issues as soon as possible.

Regarding the documentation, yes, it's mostly generated automatically, by Arturo, using Arturo's inner code structure (And yes, the tools/docgen.art script is mainly the... culprit). Regarding wrong function descriptions, since they are all picked up directly from the code as well, I guess it's just a case of me having copy-pasted the same starter code to implement a different - but similar - function. I'll definitely have to have a look into it too. :)

Thanks for pointing it out.

Of course, also thanks for your kind words. Arturo is a love child of mine and any input/feedback or contribution is more than welcome. As I said, I just had to take some time off to work on more... for-profit adventure (namely, https://deleahora.com - in case you want to have a look; btw, part of the infrastructure behind the website *is* actually written in Arturo (and Ruby))

Have a great day! :)

**Reply by drkameleon** [reply]
A little heads-up after some time :)

**2 & 3)** Fixed.

**4, 5 & 6)** Fixed. (I re-generated the documentation after fixing the description.

Basically, the website generation is a two-fold process:

- Generate the actually website content using `./tools/sitegen.art`
- Use Webize (also written in Arturo) to process the final result and produce the HTML content (+CSS/JS/images):

`webize --build --at: arturo-lang.io` (and the files land in the correct folder ready for SFTP syncing)

Regarding **1)**, you do have a point and I'm quite perplexed to be honest. I've read this - or something along the lines - as good practice. Have a look at this for example: https://css-tricks.com/the-fastest-google-fonts/

But, yes, apparently you are right: the aforementioned URL returns a 404... I'll have to investigate more. Perhaps Google changed - again - something in the meantime?
